### PR TITLE
DataGrid - Separator position is incorrect on dragging a column from the group panel to the first column (T1174763)

### PR DESCRIPTION
**Pull request by pomahtri**

### CODE REVIEW

File: js/__internal/grids/grid_core/columns_resizing_reordering/m_columns_resizing_reordering.ts [hunk 1]
```diff
@@ -1231,12 +1231,16 @@ class DraggingHeaderViewController extends modules.ViewController {
...
+
+    const isColumnAfterExpandColumn = prevColumn?.command === 'expand';
+    const isLastExpandColumn = targetColumn?.command === 'expand' && nextColumn?.command !== 'expand';
 
     switch (location) {
       case 'columnChooser':
         return true;
       case 'headers':
-        return (sourceColumn && !sourceColumn.allowReordering) || (!targetColumn || !targetColumn.allowReordering) && (!prevColumn || !prevColumn.allowReordering);
+        return !isLastExpandColumn && (isColumnAfterExpandColumn || (sourceColumn && !sourceColumn.allowReordering) || (!targetColumn || !targetColumn.allowReordering) && (!prevColumn || !prevColumn.allowReordering));
```
Comment (williamvinogradov) [inline js/__internal/grids/grid_core/columns_resizing_reordering/m_columns_resizing_reordering.ts]: Let's simplify this condition and write something like this:

```typescript
const isSourceColumnAllowReordering = sourceColumn && !sourceColumn.allowReordering;
const isOtherColumnAllowReordering = !targetColumn?.allowReordering && !prevColumn?.allowReordering;
return !isLastExpandColumn && (isColumnAfterExpandColumn || isSourceColumnAllowReordering || isOtherColumnAllowReordering);
```

And I'd like to point out that use ```||``` with ```&&``` on one level - it's a bad practice 
Let's add brackets in cases like this instead:
```typescript
// bad
return a || b && c;
// good
return a || (b && c);
```

File: testing/testcafe/tests/dataGrid/columnReordering.ts [hunk 2]
```diff
@@ -129,3 +129,46 @@ test('The separator should display correctly when dragging column', async (t) =>
   allowColumnReordering: true,
   allowColumnResizing: true,
 }));
+
+test('column separator should work properly with expand columns', async (t) => {
+  const dataGrid = new DataGrid('#container');
+  const columnSeparator = dataGrid.getColumnSeparator();
+
+  await t.drag(Selector('.dx-group-panel-item').nth(0), 0, 30);
```
Comment (williamvinogradov) [inline testing/testcafe/tests/dataGrid/columnReordering.ts]: Let's not use selectors directly
Instead, we should add the getGroupPanel method to the DataGrid testing model

File: testing/testcafe/tests/dataGrid/columnReordering.ts [hunk 3]
```diff
@@ -129,3 +129,46 @@ test('The separator should display correctly when dragging column', async (t) =>
...
+
+  // todo add screenshot test once testcafe support holding while dragging
```
Comment (williamvinogradov) [inline testing/testcafe/tests/dataGrid/columnReordering.ts]: Testcafe can take a screenshot while drag-n-drop interaction is in process.
Take a look at ```testing/testcafe/helpers/mouseUpEvents.ts``` helper.
And for example, its usages in ```testing/testcafe/tests/pivotGrid/fieldPanel/dragAndDropFieldItems.ts``` tests

So, let's add a test screenshot here :)

File: testing/testcafe/tests/dataGrid/columnReordering.ts [hunk 4]
```diff
@@ -129,3 +129,46 @@ test('The separator should display correctly when dragging column', async (t) =>
...
+
+  await t
+    .expect(await columnSeparator.getX())
+    .eql(5); // padding before first column
+
+  await t.debug();
```
Comment (williamvinogradov) [inline testing/testcafe/tests/dataGrid/columnReordering.ts]: Let's remove this debug :)
```suggestion
```

**Comment by pomahtri** [comment]
Commits with refactoring for this PR: https://github.com/DevExpress/DevExtreme/pull/25177

File: packages/devextreme/js/__internal/grids/grid_core/columns_resizing_reordering/m_columns_resizing_reordering.ts [hunk 5]
```diff
@@ -1222,21 +1222,56 @@ class DraggingHeaderViewController extends modules.ViewController {
...
+        if (isFirstExpandColumn) {
+          if (!this.isCustomGroupColumnPosition) {
+            while (columns[point.columnIndex]?.command === 'expand') {
+              point.columnIndex += 1;
+            }
+          }
+          return false;
+        }
+
+        return isColumnAfterExpandColumn || sourceColumnReorderingDisabled || otherColumnReorderingDisabled;
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/columns_resizing_reordering/m_columns_resizing_reordering.ts]: Let's reduce complexity here:

```typescript
if (!isFirstExpandColumn) {
  return isColumnAfterExpandColumn
    || sourceColumnReorderingDisabled
    || otherColumnReorderingDisabled;
}

if (this.isCustomGroupColumnPosition) {
  return false;
}

while (columns[point.columnIndex]?.command === 'expand') {
  point.columnIndex += 1;
}

return false;
```

or

```typescript
switch (true) {
  case !isFirstExpandColumn:
    return isColumnAfterExpandColumn
      || sourceColumnReorderingDisabled
      || otherColumnReorderingDisabled;
  case this.isCustomGroupColumnPosition:
    return false;
  default:
    while (columns[point.columnIndex]?.command === 'expand') {
      point.columnIndex += 1;
    }
    return false;
}
```

**Review by williamvinogradov — APPROVED** [review]

**Pull request merged**